nixos/modules/services/cluster/k3s/default.nix

```diff
@@ -190,5 +190,25 @@ in
         );
       };
     };
+
+    # K3s doesn't stop containers when service is stopped.
+    # On host shutdown, containers are still running and needs to be stopped for avoiding data corruption.
+    systemd.services.killall-k3s = {
```

Comment:
Thanks for bringing awareness of the Calico/eBFP use case. I eventually want to use Calico for myself.

> I would suggest we make this possible to use, but optional so that someone could inject their own cleanup script if desired.

The script is handling the default use case for K3s. (Which currently suffers from data loss and corruption if containers aren't gracefully stopped on shutdown.)

Users can already overwrite or disable the script by overwriting `systemd.services.killall-k3s` rule. For example:
> `systemd.services.killall-k3s = lib.mkForce { };`

But this could be documented. Or if you want it to be a flag, it should be a reversed flag, for disabling this. Since default configuration already require this as is.

A release note for this addition likely is desirable.

For Calico (eBPF based), ideally a solution for it should be documented and codified too.

I happen to have 2 other abstractions I use for K3s (`k3s-node` (local to node) or `k3s-cluster` (global) in NixOS that is missing from nixpkgs. I wonder, the correct way of approaching this. One abstraction covers the configurations for the node (which configures node services, such as longhorn, NFS, etc). The other for the cluster (which configures `k3s-node` in several nodes automatically). Anything else that is required for a functioning Kubernetes cluster is configured that way. I need to think about this, I am not ready to upstream this as it needs clean-up. My queue of to do things for K3s is quite long by now.

Are you using Calico?
What do you think the actual solution for Calico is going to be?

Reply:
In addition to flannel, the killall script also expects the default location of the data/storage directories, the default integrated containerd runtime and possibly more default values. I agree that running it on shutdown should be optional.

Reply:
> The script is handling the default use case for K3s. (Which currently suffers from data loss and corruption if containers aren't gracefully stopped on shutdown.)
>
> Users can already overwrite or disable the script by overwriting `systemd.services.killall-k3s` rule. For example:
> 
> > `systemd.services.killall-k3s = lib.mkForce { };`
> 
> But this could be documented. Or if you want it to be a flag, it should be a reversed flag, for disabling this. Since default configuration already require this to avoid data loss as is.
> 
> A release note for this addition likely is desirable. Since this is a meaningful change. And should be carefully considered.

I would prefer a flag and off by default, but would be fine with on by default or otherwise documented. The killall script uses `kill -9`/SIGKILL on your processes so I am not sure I would call it graceful for shutdown. My current use case uses kured to drain the nodes and perform reboots as needed after system updates and I manually drain if I am directly working on a node and need to reboot. 

> For Calico (eBPF based), ideally a solution for it should be documented and codified too.
> 
> Are you using Calico? What do you think the actual solution for Calico is going to be?

I am not using Calico, just know it would not be covered by the script as written and made the assumption that if it were important to a user they would at least have the documentation on how disable the existing script or otherwise inject their own.

> I happen to have 2 other abstractions I use for K3s, `k3s-node` (local to node) and `k3s-cluster` (global to cluster), as NixOS module that is missing from nixpkgs.
> 
>     * `k3s-node` configures node services, such as longhorn, NFS, etc.
> 
>     * `k3s-cluster` configures `k3s-node` in each node/host automatically (as defined in cluster configuration).
> 
> 
> Any service or feature used in Kubernetes cluster I configure that way.
> 
> I wonder, the correct way of approaching this (Calico configuration). (Where should this code go.)
> 
> I am not ready to upstream my abstractions yet because it needs clean-up. My queue of to do things for K3s is quite long by now. Just another item... But still, I think, this likely a good direction to go. I just need to polish it so we can debate how acceptable this is for nixpkgs. By that logic, I keep `services.k3s` lean, mimicking upstream configuration patterns. And the services complexity goes to `k3s-node`. This was the result of a debate I had with @euank long ago. This pattern, seems a good pattern. We should be sharing these advances we have. Others will likely find it useful and can also extend it to their needs and we can benefit from mutual contributions/expertise.
> 
> (These abstractions being obviously optional.)

Those would be cool as optional abstractions. My current use case still lets flux handle the cluster after k3s is setup so while it may not apply to me, I do like it.

Reply:
> I would prefer a flag and off by default, but would be fine with on by default or otherwise documented. The killall script uses kill -9/SIGKILL on your processes so I am not sure I would call it graceful for shutdown. My current use case uses kured to drain the nodes and perform reboots as needed after system updates and I manually drain if I am directly working on a node and need to reboot.

Then, killall script is not what we need here. IMO we should codify the ideal solution for avoiding data loss. Gracefully stopping containers before a shutdown/reboot seems reasonable for a default behavior. At least, I don't know when that isn't ideal?

> I am not using Calico, just know it would not be covered by the script as written and made the assumption that if it were important to a user they would at least have the documentation on how disable the existing script or otherwise inject their own.

Right. Current proposed solution doesn't seem ideal as it is then.

> Those would be cool as optional abstractions. My current use case still lets flux handle the cluster after k3s is setup so while it may not apply to me, I do like it.

I also use Flux, but it goes beyond that, because Flux is installed inside the cluster. I added those abstractions for doing extra configuration that is at host level. For example, adding extra packages to system or configuring NixOS services that are required for provisioning certain services. But it can also automate the injection of Flux (or ArgoCD, or anything else) for example.

Again, to clarify, It's discrete from `services.k3s` (which mimics upstream implementation). I likely should share some code for it, so it makes sense. But basically you could enable anything like `calico` and have it properly configured by default. Or any service abstraction. Goal would to codify how that works. Then, users can just enable it and have it properly configured by default. This obviously needs a lot of debate.

For example, I did not foresee the consequences that you guys are pointed out here. Too much optimism on my part. And my solution actually came pretty short. But that can also be a positive thing once we can benefit from each other expertise, as I said earlier. So that in the end we end up with something more solid than we would in our own.

Reply:
> My current use case uses kured to drain the nodes and perform reboots as needed after system updates and I manually drain if I am directly working on a node and need to reboot.

I think, we should somehow automate this (most users aren't even aware of this being necessary). Which commands are you using currently? Anything special? Thoughts?

Reply:
> Which commands are you using currently? Anything special? Thoughts?

Automated reboots are handled by [a systemd timer check](https://github.com/wrmilling/nixos-configuration/blob/main/hosts/common/optional/reboot-required.nix) which will create a file at `/var/run/reboot-required` if a reboot is needed and [kured config](https://github.com/wrmilling/k3s-gitops/blob/main/kube-system/kured/kured.yaml#L47) checking that file on a schedule and rebooting via systemd if needed

Manual reboots, I run `kubectl drain $node --ignore-daemonsets --delete-emptydir-data` which will cordon and drain the node before rebooting via SSH. 

I don't mind having the default upstream provided killall as an optional addon for those that may want to use it, just comes down to documentation so that users know its limitations and possible examples for alternative solutions. I am not sure the main module needs to catch everyone's use case, there can be so many combinations of services and edge cases to try to catch.

Reply:
I just made a very related [comment](https://github.com/NixOS/nixpkgs/issues/255783#issuecomment-2113157680) in #255783, however, I haven't really noticed data corruption so far. Do you have a setup where corruption on node shut down can be reproduced reliably and can share the config? I would like to check if the combination of graceful node shutdown + patched local-storage-provisioner prevents data corruption.

Reply:
> I don't mind having the default upstream provided killall as an optional addon for those that may want to use it, just comes down to documentation so that users know its limitations and possible examples for alternative solutions.

As it is now, it seems `killall` script would not be useful. Because it is not a graceful shutdown. And it would be already equal to what already is happening.

> I am not sure the main module needs to catch everyone's use case, there can be so many combinations of services and edge cases to try to catch.

The other abstractions I talked (k3s-node, k3s-cluster) about are new abstractions, new NixOS modules, for facilitating the configuration of services, it's unrelated to current k3s module. No changes to current K3s module would happen. I'm not advocating expansion of K3s module for services. Also, I did not properly convey how it works. Only commented because the Calico solution likely should be part of that and not the K3s module. (Considering the layers of abstractions.) But that is another topic unrelated to this.

The containers not going through a graceful shutdown on host shutdown, I think, could still be solved at the NixOS module because it is a fundamental function that seems to be causing the data corruption.

Atm I no longer plan of tackling this problem, since the proposed solution came short and was naive. And I don't have a proper solution atm.

Reply:
> I just made a very related [comment](https://github.com/NixOS/nixpkgs/issues/255783#issuecomment-2113157680) in #255783, however, I haven't really noticed data corruption so far.

Thanks.

> Do you have a setup where corruption on node shut down can be reproduced reliably and can share the config?

Unfortunately I don't have a reproducible solution for that.

> I would like to check if the combination of graceful node shutdown + patched local-storage-provisioner prevents data corruption.

I think, the graceful shutdown of containers on host shutdown makes sense to avoid data corruption of containers. I was trying to guarantee that. But even that, has to be tested too.

For now, I'll just close this naive PR.